packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart

```diff
@@ -0,0 +1,90 @@
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+import 'package:neon/blocs.dart';
+import 'package:neon/models.dart';
+import 'package:nextcloud/dashboard.dart' as dashboard;
+import 'package:rxdart/rxdart.dart';
+
+abstract class DashboardBlocEvents {}
+
+abstract class DashboardBlocStates {
+  BehaviorSubject<Result<Map<dashboard.Widget, dashboard.WidgetItems?>>> get widgets;
+}
+
+class DashboardBloc extends InteractiveBloc implements DashboardBlocEvents, DashboardBlocStates {
+  DashboardBloc(this._account) {
+    unawaited(refresh());
+
+    _timer = TimerBloc().registerTimer(const Duration(seconds: 30), refresh);
+  }
+
+  final Account _account;
+  late final NeonTimer _timer;
+
+  @override
+  BehaviorSubject<Result<Map<dashboard.Widget, dashboard.WidgetItems?>>> widgets = BehaviorSubject();
+
+  @override
+  void dispose() {
+    _timer.cancel();
+    unawaited(widgets.close());
+    super.dispose();
+  }
+
+  @override
+  Future<void> refresh() async {
+    widgets.add(widgets.valueOrNull?.asLoading() ?? Result.loading());
+
+    try {
+      final widgets = <dashboard.Widget, dashboard.WidgetItems?>{};
+      final v1Widgets = <String>[];
+      final v2Widgets = <String>[];
+
+      final response = await _account.client.dashboard.dashboardApi.getWidgets();
+
+      for (final widget in response.body.ocs.data.values) {
+        if (widget.itemApiVersions.contains(2)) {
+          v2Widgets.add(widget.id);
+          widgets[widget] = null;
+        } else if (widget.itemApiVersions.contains(1)) {
+          v1Widgets.add(widget.id);
+          widgets[widget] = null;
+        } else {
+          debugPrint('Widget supports none of the API versions: ${widget.id}');
+        }
+      }
+
+      if (v1Widgets.isNotEmpty) {
+        debugPrint('Loading v1 widgets: ${v1Widgets.join(', ')}');
+
+        final response = await _account.client.dashboard.dashboardApi.getWidgetItems(widgets: v1Widgets);
+        for (final entry in response.body.ocs.data.entries) {
+          widgets[widgets.keys.singleWhere((final widget) => widget.id == entry.key)] = dashboard.WidgetItems(
+            (final b) => b
+              ..items = entry.value.toBuilder()
+              ..emptyContentMessage = ''
+              ..halfEmptyContentMessage = '',
+          );
+        }
+      }
+
+      if (v2Widgets.isNotEmpty) {
+        debugPrint('Loading v2 widgets: ${v2Widgets.join(', ')}');
+
+        final response = await _account.client.dashboard.dashboardApi.getWidgetItemsV2(widgets: v2Widgets);
+        for (final entry in response.body.ocs.data.entries) {
+          widgets[widgets.keys.singleWhere((final widget) => widget.id == entry.key)] = entry.value;
```

Comment:
This is really ugly.
I'd rather have the id as the key of the map.

Reply:
But we need to have it as a map of widgets and their respective items. I improved it by introducing an extension for finding the widgets by id.

Reply:
Maybe a map with the id as the key and a record of widget and entry as value?
Or just have two normal maps and merge them before returning.

In the worst case the current code takes `n²` while having two maps and combining them later is in `2n`. I don't expect us to notice any performance benefit as we the number of widgets will be relatively small but lets make it cleaner from the beginning.

Reply:
Can you explain a little more how you want to use two maps and then merge them later? I'm not sure I understand what you want (although I understand why).

Reply:
I must have mixed up something in an earlier review as I also can't see the change I wanted anymore.

Reply:
I changed it a bit already